### ISSUE TITLE
drift implementation potentially deviate from intention

### ISSUE THREAD

**Issue by Paul-St-Young** [QMCPACK/qmcpack#544]
The scaling factor of the drift vector, used in importance sampling, may deviate from the intended implementation of "eq. (34) *tau in Umrigar, JCP 99, 2865 (1993)". The deviation happens whenever the wavefunction gradient has a non-zero complex component. Example code snippet from `DriftOperators.h`:

```c++
template<class T, class TG, unsigned D>
inline T getDriftScale(T tau, const ParticleAttrib<TinyVector<std::complex<TG>,D> >& ga)
{ 
  T vsq=Dot(ga,ga);
  return (vsq<std::numeric_limits<T>::epsilon())? tau:((-1.0+std::sqrt(1.0+2.0*tau*vsq))/vsq);
}
```
In the above, `vsq` may become negative if `ga` has a significant complex component. I think the `vsq` in Umrigar, JCP 99, 2865 (1993) is intended to be a non-negative real number. Thus implantation deviates from intention.

A realistic occurrence of the aforementioned situation can be reproduced using the attached nexus script (solidh.zip). In that example, the symptom of the bug is as follows. The variance in  the "vmc" folder will be > 2 ha^2 near the configuration I have chosen (using a random seed). This is a lot higher than the optimized variance ~0.4 ha^2 in the "opt" folder.

The cause of the bug is two-fold:
1.  The initial random arrangement of electrons is near a node of the many-body wavefunction causing large complex component in the wavefunction gradient.
2. drift vector implementation deviates from intention at this configuration.

Cause 1 can be verified by running a wavefunction test (wftest.zip) at the initial configuration. The resultant laplacian values are shown in the figure below. 
![laplacian](https://user-images.githubusercontent.com/8060441/33735396-c38f88f6-db54-11e7-9d16-0423a5724e4b.png)

The same symptom was observed in #514, albeit the cause was mis-assigned to `rcut`.

[solidh.zip](https://github.com/QMCPACK/qmcpack/files/1540429/solidh.zip)
[wftest.zip](https://github.com/QMCPACK/qmcpack/files/1540375/wftest.zip)

**Comment by rcclay** [comment]
After some chats, here's what's happening:
VMC/DMC/RMC should be using |Psi_T| as the guiding function, which means that the wavefunction force used to compute the drift should be calculated as the real part of grad( log(Psi_T) ).  Umrigar drift scaling should be computed with respect to this object, basically by computing the square magnitude of the REAL wavefunction force.  

Here's what the code does instead (see src/QMCDrivers/DriftOperators.h, in particular, getScaledDrift()):
1.) Compute G=grad( log(Psi_T) )
2.) Compute initial vsq = Re(G dot G) (you read that right.  No complex conjugation)
3.) Compute scale factor from Umrigar's scheme, set it equal to vsq.
4.) Calculate vsq*G, and set the drift equal to the real part of this.  

This should give the right answer for a real wavefunction, but definitely not for a complex wavefunction.  The correct thing to do is to replace 1.) with G=Re(grad( log(Psi_T) )), or to ensure that this step is done first at the driver level before calling these functions.

**Comment by rcclay** [comment]
I'll push a fix of this later, but here's roughly what needs to be done:

1.)  Make sure the wavefunction force is real before evaluating umrigar drift scaling.  I'm flushing this out right now, but fixing getScaledDrift should fix nearly all the issues in the code.  DMCNonLocalUpdate and DMC_CUDA drivers call getDriftScale instead and compute the drift by hand, so the fix for these can be to ensure that G is real before calling getDriftScale.  
2.)  DriftOperators.h needs to be worked through a bit more carefully to see what functions are being used in the code or will likely need to be used in the code.  Any dependency found relying on a gradient needs to be checked.  (Optional:  remove any function not being used by the code).

**Comment by rcclay** [comment]
On the plus side, this bug goes to zero as tau->0, and should be extrapolated away with careful timestep convergence tests and extrapolations, so don't everyone panic.  In this context, it might be more of a timestep error enhancer than mitigator.

**Comment by Paul-St-Young** [comment]
Agreed. I also vouch for the removal of the ternary `(vsq<std::numeric_limits<T>::epsilon())? tau:`.
It hides the bug.

**Comment by jtkrogel** [comment]
This is interesting since I've generally found the complex code to have smaller timestep errors than the real one, even for real-valued wavefunctions.

**Comment by jngkim** [comment]
I don't think we have done enough with wbyw until Paul is seriously looking
into it.
So, glad that these subtle bugs are being crushed.

**Comment by Paul-St-Young** [comment]
@jtkrogel you reminded another point I wanted to bring up. The `a` parameter in Umrigar's eq. (34) is set to 1.0 in the code. The large-drift limit Umrigar's drift is sqrt(2*tau/a), where `tau` is time step in Hartree atomic units (a.u.). The more physical (via discussion with David Ceperley) upper limit should be the RMSD of diffusion length, namely sqrt(tau/m) in each spatial dimension . `m` is the mass of the particle in a.u.. The Umrigar upper-limit is larger than the Ceperley upper-limit.

If the wavefunction gradient has a small imaginary component, then it would reduce the drift. This may explain the smaller timestep when using the complex code path.

**Comment by Paul-St-Young** [comment]
@jngkim the bug was reproduced with the pbyp driver in solidh.zip

**Comment by jngkim** [comment]
Yes, I know.

**Comment by ye-luo** [comment]
@Paul-St-Young I'm looking at the Umrigar's scheme of Green function modification and I will pay more attention to the detail you mentioned.

@rcclay I checked the math.
Based on the assumption VMC/DMC/RMC using |Psi_T| as the guiding function.
G=grad(log(|Psi_T|) =Re(grad(log(Psi_T)).
So at the driver level, we should always deal with Real. Namely, evalGrad and ratioGrad should return RealType from TrialWavefunction for PbyP moves.

In general, we will also have a great simplification of DriftOperators.h.

**Comment by prckent** [comment]
Great to see this progress.

By the time of @ye-luo 's Green function investigation, we should have these points carefully written up in the manual so that we don't have to rediscover them a few years from now...

**Comment by Paul-St-Young** [comment]
Dear all, after further discussion with Prof. David Ceperley, I believe we must re-examine the ramifications of this bug in DMC carefully. If the wavefunction gradient has a small complex component, then the current DMC drift is WRONG.

I must also reverse my opinion on the ternary `(vsq<std::numeric_limits<T>::epsilon())? tau:`. It saves us when the imaginary part of grad_psi_over_psi exceeds its real part. Namely, this ternary resets the DMC drift to the naive drift, which is correct (albeit with a rather large time-step error). However, when the imaginary part is small, the DMC drift is WRONG and we apply the WRONG Green's function, resulting in a systematic bias. A few details might limit the size of this bias, for example Metropolis-Hasting rejection, which enforces continuity. Nevertheless, the magnitude of this bias needs to be assessed.

**Comment by ye-luo** [comment]
@rcclay and I agree that the Umrigar scheme should be applied based on the G.real(). This is clearly a bug we have.

The ternary is intended to avoid numerical issue but in fact it also save us when the complex part is larger than the real part. Once the bug is removed, the ternary will do the work it is supposed to do.

**Comment by rcclay** [comment]
@Paul-St-Young  I agree that we should try to assess the size of the bias.  However, whatever bias there is, I suspect it is small.  Otherwise, it would have been flagged in the ctest after this fix was introduced.  Here's why I think its not that big of a deal in practice:

So, what the naive drift scaling should do is compute the drift term as tau*Re(F), where F=grad(log(psi))  (There might be a factor of two in here, whatever.  Shooting from the hip). One way you can think about what Umrigar drift scaling does is to add a scale factor "v" depending on |F| such that drift = v*Re(F).  When |F| is smallish, v will reduce to tau, yielding the naiive value.  When |F| is large, the magnitude of the drift reduces to sqrt(2*tau).  So this scheme is a magnitude limiter on the drift term.  

So what does this bug do?  The drift computed by QMCPACK is real and proportional to the correct drift, but is off by a scale factor based on the incorrect computation of v.  Notice that there's no pollution from the imaginary part of the trail wavefunction gradient, except for the scale factor.  When will the scale factor matter?  1.) large |F| (like near nodes), and 2.) large tau.  Fortunately, provided one drives tau to small enough values, v is exceedingly likely to be equal to tau, and the naiive drift will prevail.  Is this the correct DMC algorithm with the correct DMC drift?  Technically no.  However, you can probably convince yourself that this bug is a timestep error, and will recover the exact algorithm in the limit as tau->0.  Provided normal users do their usual timestep extrapolations and use small enough timesteps, I think the answers will be OK.

**Comment by Paul-St-Young** [comment]
@rcclay I would agree with your analysis if we were actually calculating the Umrigar scaling factor in `DriftOperators.h`. The unfortunate fact is that the variable `sc` does NOT hold the Umrigar scaling factor of eq. (34). Rather `sc` holds `tau*eq (34)`. What we calculate reduces to the naive drift of a different time step as `tau->0`.

`sc = tau+2/3*tau^2*v^2+O(tau^3*v^4)`. A contaminated `v` contaminates the time step of the drift part of the Green's function. Of course, the contamination is most severe near a node, which we are prohibited to visit due to rejection. However, I must also point out that the bug underestimates `v`. It is most dangerous when the imaginary part of the wavefunction gradient is comparable but still smaller than its real part. There we have the most serious contamination of the drift Green's function AND we lose the protection of the drift, which is supposed to push us away from the node. We are save by the ternary only when `vsq<0`.

**Comment by Paul-St-Young** [comment]
OK `sc` is still correct if `tau` is "very small", but the definition of small depends on the magnitude of `v`. I think the correct definition of "very small" in this case is `tau << 3/(2*v^2)`.

**Comment by Paul-St-Young** [comment]
@rcclay I must apologize, I was not thinking straight at the end of last night. The drift vector does reduce to its naive value with the same time step. I did my analysis using `eq (34)` and incorrectly transferred some of my conclusions to `tau * eq (34)`, which is what we have.

To match symbols with the code, define `drift = sc * v`, where `v` is grad(log(|psi|)). `vsq=v \dot v`.

I agree with your analysis now, and below is a plot to accompany your explanation. The bug has the effect of setting `sc` higher than it should be. Thus whenever the wavefunction gradient has an imaginary part, the drift is a bit closer to the naive drift than intended. One thing I noticed from the plot is that the error in `sc` becomes insensitive to `vsq` as `vsq` becomes large. That might have saved us a bit. When `vsq` is small, the drift is close to naive anyways, so we are also OK there.

![drift_bug](https://user-images.githubusercontent.com/8060441/33797373-4a3faed0-dccc-11e7-8b73-6ec96d09402c.png)
